### DISCUSSION TITLE
写了一个高并发、高可用的营销活动，欢迎大家沟通交流

### DISCUSSION THREAD

**Discussion by wenbochang888** [just-talks/tech-talks#66] [3.分享创造]
1. 减库存。将奖品的库存信息提前缓存到redis里面，比如奖品100个缓存到redis里面。如果有100W人来抢100个奖品，最终也只有100个人通过redis的校验
2. 根据业务场景，如果不是必中奖。在减库存之前，做一个随机数。如果在随机数之外，直接返回”奖品被抢完“，限制大部分流量进入到redis减库存
3. 放弃重试 失败重试会影响系统性能，重试次数越多，对系统性能的影响越大。
4. 防止奖品超发 一般我们会通过乐观锁，悲观锁，分布式锁来解决。其中乐观锁的效率是最高的。

[核心的代码](https://github.com/wenbochang888/marketing/blob/14fc64e22e37e9f2b7d0de6ab606f35a63ca49a7/src/main/java/com/wenbo/marketing/service/MktActivityService.java#L81)在这一块，欢迎大家沟通交流，有没有更好的一些优化手段了 

[源码](https://github.com/wenbochang888/marketing)
[体验地址](http://101.33.233.214:8989/)

**Comment by adoyle-h** [comment]
几点建议：

1. demo 地址打不开。
2. 你需要提供 benchmark 来证明你说的高并发。
3. 你这个项目代码与高可用无关。
4. 抽象项目代码，把它写成一个框架，提供开箱即用的项目 example。
5. 思考一个问题： 这个开源项目，它能直接用到别人的生产环境使用吗？目前我认为不能。我觉得它只是一个 demo，不是一个工具。
6. 如果只是想分享这类心得经验，可以再梳理一下写成文章，发表到知乎或者别的地方，会有更好的传播和讨论。开源在 Github 应该达不到你想要的效果。

**Reply by wenbochang888** [reply]
感谢建议
1. demo 地址 http://101.33.233.214:8989/，我刚才试了下，可以打开，你可以试试把代理关了
2. benchmark我测过，由于服务器是小水管，一压 cpu就起来了，但整体项目负载，mysql等负载还比较低
3. 有关的，主要是思路。比如缓存，比如通过业务手段拦截流量，甚至通过mq异步减库存等等
4 - 6 其实他并不是一个开源项目，只是提供一种思路，一种心得体会，给大家在写业务之余，提供一些营销系统常见的一些优化手段

如果对整体高并发思路，有更好的建议，欢迎交流沟通

**Reply by wenbochang888** [reply]
这其实并不是一个完成的 开箱可用的 成熟的项目，他只是我花了2天写出来的一个demo。提供给大家的一个思路，以此为契机，大家可以展开讨论，交流想法
